Release 5.0

dougwilson opened the pull request:
This is a tracking issue for release 5.0.

**_5.0.0-beta.1 published on npm_**

I am trying to give better visibility for upcoming changes and so am trying out making a PR for a release here, pulling from the official next release branch into 4.x. This allows for the current pending changes to be easily visible.

**Please keep feature requests in their own issues**

I'm also leaving this PR unlocked so people can make comments/etc. and we'll see how it goes :) If you want to make a comment on a particular change, please make the comment in the "Files changed" tab so comments are not lost during a rebase (they will occur frequently has 4.x progresses).

The goal of Express 5 is to be API tweaks & the removal of all code from the Express repository, moving into components in the pillarjs project (https://github.com/pillarjs), providing at least basic support for promise-returning handlers and complete HTTP/2 functionality. Express 5 would become a "view into pillarjs" and would be an arrangement of these components.

List of changes for release:
- [x] Add support for Promises in all handlers #2259 (though `router@2.0.0-alpha.1`)
- [x] Add support for Promises in `app.param` (though `router@2.0.0-beta.2`)
- [x] Bring back `app.router` which is just generally useful for directly calling to the router
  - [ ] Make sure 3.x-style `app.use(app.router)` does not explode
- [ ] ~Fix views to resolve paths async #2653~
- [x] Make `query parser` option default to `'simple'` #3361
- [x] Make `bodyParser.urlencoded` default to `'simple'` (though `body-parser@2.0.0-beta.1`)
- [x] Make `req.host` actually return the host #2179
- [x] Make `req.query` a getter instead of added by middleware #2215
- [x] Make `res.render` always async callback, to hide sync view engines #2668
- [ ] ~Make `res.render` use the default engine even when an extension was found #2708~
- [ ] ~Make `res.sendFile` use the `"etag"` application setting #2294~
- [x] Make `res.status` throw on invalid argument type #4212 #2795 #2797 #3111 #3137 #3143
- [ ] ~Make `express.static` use the `"etag"` application setting #2317~
- [x] New path matching syntax #2057 #2173 #4321 (though `router@2.0.0-beta.1`)
- [ ] ~Provide separate locals and options to view engine #2648~
- [x] Remove Express 3.x middleware error stubs #3217
- [x] Use `mime-types` instead of `mime`
- [x] Use `path-is-absolute` module for absolute path detection #2620 

List of deprecated things removed:
- [x] Leading `:` in `name` for `app.param(name, fn)`
- [x] `app.del`
- [x] `app.param(fn)`
- [x] `req.acceptsCharset`
- [x] `req.acceptsEncoding`
- [x] `req.acceptsLanguage`
- [x] `res.json(obj, status)`
- [x] `res.json(status, obj)` #2939
- [x] `res.jsonp(obj, status)`
- [x] `res.jsonp(status, obj)` #2940
- [x] `req.param()`
- [x] `res.redirect(url, status)` #2941
- [x] `res.send(body, status)`
- [x] `res.send(status)`
- [x] `res.send(status, body)` #2942
- [x] `res.sendfile`
- [x] `res.vary()` (i.e. no arguments) #2943

List of things to refactor (still provided by `express`):
- [ ] ~Better `res.cookie` and general cookie reading/setting (new `keygrip` and `cookies` modules; new `req.cookies` or such)~
- [ ] ~Move prototype properties to own repo #2432~
- [x] Move `Router` to it's own repo #2411

**Testing this release**

If you want to try out this release, you can install it with the following command:

``` bash
$ npm install expressjs/express#5.0
```

Owners/collaborators: please do not merge this PR :)

dougwilson committed:
Change req.host to return host

dougwilson committed:
Remove res.send(body, status) signature

dougwilson committed:
Remove res.json(obj, status) signature

dougwilson committed:
Remove res.jsonp(obj, status) signature

dougwilson committed:
Remove app.router error message

dougwilson committed:
Change req.query to a getter

dougwilson committed:
Refactor away init middleware

dougwilson committed:
Add app.router reference back

jonathanong commented:
here's some mostly refactor-based changes i'd like to see. AFAIK you've got this covered @dougwilson, but let me know if you need help:
- refactor the router to a separate repository
- refactor the view system to a separate repository. maybe [templation](https://github.com/expressjs/templation), maybe something else more backwards compatible
- finish next versions of `cookies` and `keygrip` and use them in Express
- finish refactoring `body-parser` into the `request-body` module before including it here
- refactor the router to use [path-match](https://github.com/expressjs/path-match) or something else between `path-to-regexp` and the router itself
- refactor [proxy-addr](https://github.com/expressjs/proxy-addr). i just emailed npm to ask for the `trust` package

dougwilson commented:
All great stuff @jonathanong :) I added all but the `body-parser` one up there, because I really want to add that in a 4.x release, but time-dependent :D

ilanbiala commented:
How close is this to a prerelease? I just want to know if you think it's stable enough to start playing around with for future projects.

dougwilson commented:
So you're always welcome to play around with `npm install visionmedia/express#5.0` (and you can add `"express": "visionmedia/express#5.0"` to your `package.json` if you really wish). If you are looking for something published to `npm`, just ask :) No one has really ask for it yet is all. Would you like an alpha or something published to npm? I can do it right after 4.10 (which is due something this week) so it'll include 4.10 stuff.

rlidwka commented:
> If you are looking for something published to npm

I'm looking for the router to be published to npm for the looong time now. :)

dougwilson commented:
@rlidwka the docs are just about complete and I was just porting the recent changes to the router in here over yesterday :) I also really want it, haha

Fishrock123 commented:
:+1:

ilanbiala commented:
Can someone link me to the docs for the new router?

dougwilson commented:
@ilanbiala the 5.x router is the same as the 4.x router, so the main docs can be found at http://expressjs.com/4x/api.html#router ; the documentation talked about above is basically a rewrite of them to live with the extracted router.

dougwilson committed:
Remove app.del

dougwilson committed:
Remove utils.contentDisposition

dougwilson committed:
Remove req.acceptsCharset

dougwilson committed:
Remove acceptsEncoding

dougwilson committed:
Remove req.acceptsLanguage

dougwilson committed:
Remove res.sendfile

dougwilson committed:
Remove res.send(status) signature

dougwilson committed:
5.0.0-alpha.1

dougwilson commented:
Everyone on this thread: The first Express 5.0 alpha has been released to npm!

``` sh
$ npm install express@5.0.0-alpha.1
```

hacksparrow commented:
Is there an approx ETA for 5.0?

Fishrock123 commented:
> Is there an approx ETA for 5.0?

Yep; whenever Pillarjs is complete enough to make express one file. XD

hacksparrow commented:
Haha, is it getting close?

ilanbiala commented:
@Fishrock123 What's pillarjs?

dougwilson commented:
@hacksparrow there is no real ETA just yet, mostly because I don't want to have to commit to a date :) There have been people asking on IRC/Twitter/here about getting something on `npm` from this branch, so the alpha 1 is that :) What do you think about making a 5x section on the website? I feel like it's probably too early, personally. Perhaps when it's beta, and definitely when it's RC, unless you think otherwise, lol.

hacksparrow commented:
@dougwilson, that's what I wanted to get a feel of. I am in agreement with "Perhaps when it's beta, and definitely when it's RC".

Fishrock123 commented:
@ilanbiala pillarjs is a collection of framework components. It's being extracted from express: https://github.com/pillarjs - more info is available in [this empirenode talk](http://youtu.be/FiHI6--wB4s?list=PL31ehRjJCA6Fh_l4jd-0neUKLUDwycZK8)

dougwilson commented:
@hacksparrow so my general feel was alpha Nov 1, beta Dec 1, RC Jan 1, release 1 week after RC unless issues brought up.

hacksparrow commented:
@dougwilson sounds good!

Fishrock123 commented:
@dougwilson are we still keeping static in here?

dougwilson commented:
Probably.

Twipped commented:
Is this still up to date?

dougwilson commented:
Yep.

Twipped commented:
Isn't req.param() being removed?

Fishrock123 commented:
@ChiperSoft see #2450

Twipped commented:
@Fishrock123 right. is it only being deprecated and not removed in 5.0?

dougwilson commented:
So until I actually deprecate them and see how it blows over, I don't want to commit to removing them from 5.0. If there isn't enough time, they may just was well stay as deprecated.

ilanbiala commented:
Is 5.0 going to come with new features/big performance fixes, or is it just more modular?

Fishrock123 commented:
New features: a few, yes. (Performant custom prototype extensions are one, more info soon)

"Big performance fixes": what? Do elaborate, express is by no means close to slow. The router will have some increased performance though.

ilanbiala commented:
@Fishrock123 not saying it's slow, but there's always up, right? :)

dougwilson commented:
The answer is we are not aware of any performance issues, so no, there wont be any changes there. If there is something, we wouldn't really wait for a major to fix anyhow. During the 4.x series many performance issues were addressed, as an example.

There is even a slight router performance change going into 4.11.

kuanslove commented:
@dougwilson  Hi, Doug, thanks for reply. As beginner in Node and fall in love with Express.js for more than one year,   I am really concerning about its future, will it still be supported as independent light-weight easy-to-start  core-robust-but-not-too-fancy-feature project that every new beginner can quickly start with, OR will it be put a lot of feature into and gradually merged into StrongLoop to make it like Django in Python? From the talk above, I can tell Express is going to be more robust, but I am not sure if it is going to be Fat?

dougwilson commented:
@kuanslove As far as core Express goes (what you get with `npm install express` is not getting any fatter and only potentially slimmer.

dougwilson commented:
Basically the general sort answer to your question, @kuanslove , is Express's future is "more of the same, but more flexible".

kuanslove commented:
@dougwilson  Cool, that is great! This is the exact place which attracts me so much. Another question is about Language Trend comparison: From TJ's early post farewell to node, I start to know the Golang, I also want to hear your opinion about the comparison between these two in Web development area. Most posts about Go and Node I found talking about GO is good at speed and Concurrent, but I am not sure if those PROs is actually commonly used a lot in our daily coding life for most web projects?

dougwilson commented:
I have no experience with Go. In my day job I write in C#, and as hobbies I write in Perl (which also works with an event loop--even libuv, the underlying loop for Node.js) and JavaScript, all of which have concurrency, so there's no insight I can really give you on Go.

kuanslove commented:
@dougwilson Thanks Doug! Speaking of C# and even Asp.net, I wonder why there are a lot of people use it,  is that because it has a whole set of solution for web development? Like from IDE to Platform? I do not have any experience on Windows web dev, do you think it still a huge promising job market? I am thinking if I need to put some C#/Asp.net experience in my knowledge tree. [It is my fault, I guess this should be talked offline, what is your Twitter id]

TimothyGu commented:
Is the locals/view engine options separation being worked on? I don't see a ticket regarding that

dougwilson commented:
@TimothyGu yes, if you're referring to the discussion from `ejs` a little bit back :) Indeed, there is no ticket, mainly because I don't have any code to really show for it, but feel free to open any issues/PRs you like if you have suggestions for how `__express` and `res.render` can be improved, from your point of view :)!

dougwilson commented:
@kuanslove my email in in my GitHub profile :)

TimothyGu commented:
@dougwilson said:

> @TimothyGu yes, if you're referring to the discussion from `ejs` a little bit back :) Indeed, there is no ticket, mainly because I don't have any code to really show for it, but feel free to open any issues/PRs you like if you have suggestions for how `__express` and `res.render` can be improved, from your point of view :)!

OK cool. I'll try to find some time to do that. What I have in mind however is something that looks like `res.render('file', options, locals, cb)`, with `app.locals`, `app.viewOptions`, `res.locals`, and `res.viewOptions`.

dougwilson commented:
@TimothyGu great!

dougwilson committed:
Merge tag '4.11.1'

dougwilson committed:
docs: simplify 5.x history

ForbesLindesay commented:
@TimothyGu loop me in when you create the issue :)

jancarloviray commented:
Awesome. Any chance on including some ES6 features? generators, yield, promises, etc.?

dougwilson commented:
@jancarloviray there are already a few issues at https://github.com/strongloop/express/issues under discussion.

dougwilson committed:
Merge tag '4.11.2'

dougwilson committed:
Merge tag '4.12.0'

Fishrock123 committed:
Use path-is-absolute module for absolute path detection

closes #2620

dougwilson committed:
Merge tag '4.12.4'

aredridel commented:
Would love to see #2653 in the 5.0.0 release!

dougwilson commented:
I guess I was mistaken when I said this check list was complete :) I'm adding that PR here and making sure everything currently tagged with "5.x" is listed here (because, really, they mean 5.0.0).

jfmelo commented:
:+1:

mgenev commented:
I'm curious... with Koa and Express being so similar, is Express being developed towards ES6-7 and minimalism like Koa, is Express the robust, for-prod one and Koa the experiment etc...

dougwilson commented:
@mgenev I'm not sure specifically what you're asking, but Express and Koa are just different approaches at being a web server in Node.js; one would not become more like the other and they are going to continue down the same paths they are already on.

crandmck commented:
@dougwilson As best I can tell, the ETA for initial 5.0 release is "July 2015".  Can you provide a more specific date?  We'd like to release the docs at the same time, and we are working on them...  Just need a target date if possible.

dougwilson commented:
No problem, @crandmck :) I swore I wrote this down _somewhere_, though clearly it's not in this thread, haha. So I plan to release the beta 1 this weekend (about June 29), followed by any follow ups or RCs each weekend of July until the final release July 29. The final release is not absolute stone, though, if a user brings up a big issue that we cannot resolve by then for some reason, of course :) Hope this helps!

crandmck commented:
Thanks @dougwilson that's very helpful.  The docs might not be quite ready by this weekend, but I will talk with @hacksparrow and see how soon we can get them ready.   

My suggestion would be that for beta releases, the 4.x docs should remain the "default" docs, and then once 5.0 final is released ~7/29, we would switch to the 5.0 docs as the default replacing the 4.x docs (with 4.x API docs still available of course).  The migration guide would probably come in somewhere in there, perhaps not 'til 7/29.

dougwilson commented:
@crandmck that's no problem. We don't even really _need_ docs ready until release, and even this, probably can simply skip on having them all the way to an actual RC.

hacksparrow commented:
`app.param(function)` and `router.param(function)` were deprecated as of 4.11.0, but they seem to be working fine and dandy in 5.0.0-alpha.1. The behavior is the same only the deprecation message is gone.

Or am I missing something?

hacksparrow commented:
Same thing for `req.param(name [, defaultValue])`.

hacksparrow commented:
Are we getting rid of everything that was deprecated in 4.x?  `res.send(status, body)` is still intact (with the deprecation message, of course).

@dougwilson if it would help, please assign some of the open issues to me.

dougwilson commented:
Hi @hacksparrow , yes, all deprecated things will be removed in 5.0. Right now, the current 5.0.0-alpha.1 is based off Express 4.10.1, and thus anything deprecated after 4.10.1 has not yet been removed from the 5.0 alpha. Express 4.x is essentially locked for new features this month so I can get a 5.0 beta out based off 4.13.

hacksparrow commented:
@dougwilson I am starting preliminary work on the API docs, is there a checklist of changes in 4.13 to make it 5.0 beta? It will help me in the documentation. I may also be able to help you close some of the open items.

dougwilson committed:
Merge tag '4.13.1'

dougwilson committed:
Remove ":" stripping from app.param()

dougwilson committed:
Remove app.param(fn) signature

dougwilson committed:
Remove utils.flatten

dougwilson committed:
Remove un-used depd import in router

dougwilson committed:
Remove req.param()

dougwilson commented:
Hi @hacksparrow , the check list at the very top should be up-to-date now.

dougwilson committed:
Use router module for routing

closes #2411

dougwilson committed:
Make res.render callback is always async, even for sync view engines

closes #2668

dougwilson committed:
5.0.0-alpha.2

dougwilson commented:
5.0.0-alpha.2 of Express has just been published to npm. You can install with `npm install express@5`.

hacksparrow commented:
@dougwilson that's great!

jinmatt commented:
:+1:

mgenev commented:
@dougwilson 

> I'm not sure specifically what you're asking, but Express and Koa are just different approaches at being a web server in Node.js; one would not become more like the other and they are going to continue down the same paths they are already on.

Forgive my ignorance, but what is so different? They seem close to identical to me...

amitmtrn commented:
@dougwilson apperently you can't `npm install express@5` (you can although `npm install express@5.0.0-alpha.2`)
`npm ERR! notarget No compatible version found: express@'>=5.0.0 <6.0.0'`
I installed the 5.0.0-alpha.2 version to test the [http2 issue](https://github.com/molnarg/node-http2/issues/100) #2432 and it seem that I still have the error

```
_stream_readable.js:505
    dest.end();
         ^
TypeError: undefined is not a function
    at Stream.onend (_stream_readable.js:505:10)
    at Stream.g (events.js:199:16)
    at Stream.emit (events.js:129:20)
    at _stream_readable.js:908:16
    at process._tickCallback (node.js:355:11)
```

but it's appear only on windows, on linux (ubuntu) I don't have this issue from version 4.13.1
any progress on that matter?

dougwilson commented:
Hi @amitmtrn I am able to "npm install express@5" OK, but perhaps out version of npm differ on the matter :/ As for that error, the answer is no, Express 5 alpha 2 still does not accept custom prototypes on req and res. I have not gotten to it, but PRs made against the 5.0 branch for it are welcome.

dougwilson commented:
Otherwise, you can track if an issue has been released by waiting for the issue to be closed on GitHub.

listepo commented:
Any progress?

dougwilson commented:
The current progress is represented by the GitHub milestone and the check boxes at the top of this PR :)

ariporad commented:
Hi,

I've been having some trouble figuring out what the status is on promises is. Could someone explain it (the status for express, not promises in general) for me? Thanks!

dougwilson commented:
Hi @ariporad , the check box is not checked in the list at the top, thus not done. It is not yet in progress, either.

ariporad commented:
@dougwilson: If I'd like to help implement it, what should I do. I'm a little lost with pillarjs/express/etc.

Thanks!

dougwilson commented:
Hi @ariporad , you would want to look at the https://github.com/pillarjs/router module, adding it to the handlers there, which is the router that backs Express 5.0 (https://github.com/strongloop/express/blob/5.0/lib/application.js#L27).

ronkorving commented:
It would be awesome if we could have a solution for #2781 in Express 5. HTTP/2 is the next big thing for webservers and I think it would be a great opportunity to offer support for it from the start. Now is the time, imho. Any guidance as to how we could get there would be much appreciated.

dougwilson committed:
Merge tag '4.13.4'

dougwilson committed:
deps: router@~1.1.3

dougwilson committed:
deps: array-flatten@2.0.0

tunniclm committed:
Remove res.json(status, obj) signature

closes #2939

tunniclm committed:
Remove res.jsonp(status, obj) signature

closes #2940

tunniclm committed:
Remove res.redirect(url, status) signature

closes #2941

tunniclm committed:
Remove res.send(status, body) signature

closes #2942

tunniclm committed:
Remove res.vary() (no arguments) signature

closes #2943

gabrielcsapo commented:
@dougwilson is there a community gitter or slack?

blakeembrey commented:
@gabrielcsapo Yes, it's mentioned in the README: https://github.com/expressjs/express#docs--community. See https://gitter.im/expressjs/express.

dougwilson committed:
Merge tag '4.14.0'

ilanbiala commented:
Any progress on beta/RC releases for 5.0?

dougwilson committed:
Merge tag '4.14.1'

dougwilson committed:
deps: router@~1.1.5

dougwilson committed:
deps: array-flatten@2.1.1

dougwilson committed:
deps: path-is-absolute@1.0.1

dougwilson committed:
5.0.0-alpha.3

wesleytodd committed:
call callback once on listen error

wesleytodd commented:
Can we remove the [warnings for the middleware](https://github.com/expressjs/express/blob/c8d9223e93ee0c08490e4840f3278314ccb221a5/lib/express.js#L70-L102) that was deprecated for `4.x`?  It seems to me that leaving the warnings for a whole major version/multiple years should be enough.  Right?

dougwilson commented:
@wesleytodd I thought the same thing the other night. I would like to remove them. Ideally people are much less confused these days, though I have no idea how to really gauge that without just removing them and then we can always add them back again in 5.x if it seems warranted. Feel free to make a PR to remove 😃

wesleytodd committed:
Remove Express 3.x middleware error stubs

closes #3217

wesleytodd commented:
Done: #3217

dougwilson committed:
Merge tag '4.15.0'

dougwilson committed:
5.0.0-alpha.4

dougwilson committed:
Merge tag '4.15.2'

dougwilson committed:
5.0.0-alpha.5

alamin3000 commented:
Guys when is next release coming? alpha 6 ... any roadmap we can refer to?  I know it is not easy to keep things consistent and release updates quickly.  But waiting too long historically had negative impact on some of the largest javascrpt frameworks (Angular?) Specially around this time when NodeJS and EcmaScript is aggressively moving forward and developers want to utilize them.   Any update would be appreciated.  I didn't find any info on your website.  I am mainly waiting for async middleware functions.

Thanks.  and please don't ask me to use Koa.

dougwilson commented:
@alamin3000 I've been waiting for Node.js to finally land http/2 in code to know what deep changes will need to be done in Express.js. There is (finally) a PR in Node.jc core to land http/2 behind a flag, and when the reviews go through on that, there will finally be an implementation to start building against here on this end.

The general roadmap is right at the top of this PR, in checkbox form even :)

alamin3000 commented:
Thanks.  So we are still waiting on NodeJS.  I see.. I thought with NodeJS version 8 everything was ready.  Really tough to start new project with Express 4, specially when we know 5 is on horizon and specially Promise based middleware is coming.  (I will need to write tons of middleware for the project).

BTW, as for roadmap, I meant timeline of releases... not feature list.

Thanks.

dougwilson commented:
Gotcha. Since Node.js does not provide a timeline of features, we would have nothibg to base ours on.

dougwilson committed:
Merge tag '4.15.4'

dougwilson committed:
Merge tag '4.15.5'

dougwilson committed:
5.0.0-alpha.6

dougwilson committed:
Remove path-to-regexp dependency

dougwilson committed:
Merge tag '4.16.2'

dougwilson committed:
deps: debug@3.1.0

dougwilson committed:
deps: router@~1.3.2

vvo commented:
> thanks to IBM

Wat?

@dougwilson is doing amazing work I saw commits recently.

alamin3000 commented:
I think the issue isn't about who is doing or not doing good job on this project.  I am sure everyone involved in this project is doing their best since it is backed by big players in the industry.  And it is indeed very important project in this community.  

However the issue is as a normal developer it is very frustrating when we can't use latest features and don't know how long to wait.  I have complained about this before that we don't even know the timeline (dates) for releases.  Even nodejs has timeline roadmap.

Honestly technologies are moving too fast, specially in ES and Node realm.  So progress in the ExpressJS is in that prospective is disappointing.  Again, from our (developers) prospective.  NOT from the team who is involved in it assuming they are doing their best.

I'm not looking for alternative to Express or even how to start working and Express 4 and then migrate to version 5.... I'm  strictly talking about/referring to Express 5.

stigok commented:
Please stop unnecessary discussion in this PR. It would be nice to be notified of *actual* updates to the progress.

dougwilson commented:
Please, I'm sure there is a better forum to have this conversation in. If you want to know what is currently happening with Express 5 at this very moment, the `router` module, which is the basis of Express 5 routing engine, is currently building up the 2.0 release, which is what Express 5.0 will be based on (PR https://github.com/pillarjs/router/pull/60). This encountered a bug in `path-to-regexp`, which got fixed and working through that upgrade for the 2.0 router (PR https://github.com/pillarjs/router/pull/42). I pinged about the state of the `path-to-regexp` PR just a couple hours ago.

alanpurple commented:
sorry for disturbs, all deleted

alamin3000 commented:
Is it even remotely possible to give update on dates when express 5 will be RC or release?. All I get is feature updates. I thought express is under node foundation incubator program and would have better management outside of just doing coding. Call me jerk or what, just an approximate time when we can expect to get released. Please.

dougwilson commented:
I'm not sure anyone is getting paid to work on this, and so it just falls under whenever there is time to do the next task. If there is a better methodology to do this or you have some tips on how to do release planning and such, please let us know! For example, how do we come up with a date to release?

alamin3000 commented:
If this is for sure, that this project is not under some kind of organization where compensation is available for developers and it is as you say whenever there is time, then I drop all my complain.  Honestly if is just another open source side project by some smart developers, then I don't expect timeline or anything.   I will delete my comments soon.

vkarpov15 commented:
@dougwilson it really sucks you're not getting paid to work on express. I'd love to help out with managing the project and figuring out how how to monetize if you're interested. So many companies, mine included, depend on express and want to see the project succeed.

mikeal commented:
I wouldn't assume that just because Doug doesn't have job working on express full-time that it's what he *wants* :)

Doug is a very accomplished developer who has a lot of options.

Just because other people want Doug to work on the thing *they* need doesn't mean that it's the best thing for Doug to be doing with his life. Open source is people, and people get to make the decisions that are going to make them the most happy and fulfilled, not what is best for you :)

If you want the release to get out quicker, contributions are welcome :smiley:

vkarpov15 commented:
Just putting a possible option out there, no pressure, but I'd appreciate if Doug spoke for what Doug wants rather than a third party. No way Express should be a full-time job, lib is not nearly complex enough to justify that, but the lib is stuck in 2014 and I'm happy to help :)

wesleytodd commented:
*NOTE: @dougwilson I think once this conversation has run it's course it would be worth just deleting all of these comments, including mine.*

For anyone wondering, there is a TON of active work happening in the express community.  If you are interested you can follow some of the more active repos (links below).  Also, feel free to check out the [discussions repo](https://github.com/expressjs/discussions/issues) for past meetings of the CTC (which admittedly has not happened in a while).  Supporting such a large community is a big job, and luckily @dougwilson is nice enough to commit a ton of time to it.  But that being said, it is an open community that is welcome to new people.  So if you have suggestions or PR's you want to open, then BRING IT :)

We all need to make sure we focus on more than just the major version number, and the "shiny new things".  The strength of express it not in it's speed of feature development, but in it's reliability for its users.  So lets make sure to focus on if it serves it users, not how fast the major versions increment.

Repos to follow:
https://github.com/expressjs/express
https://github.com/pillarjs/router
https://github.com/pillarjs/path-to-regexp
https://github.com/expressjs/session

czaarek99 committed:
Change query parser setting default to 'simple'

closes #3361
closes #3621

dougwilson committed:
Merge tag '4.16.4'

dougwilson committed:
deps: router@2.0.0-alpha.1

dougwilson committed:
tests: add router promise tests

dougwilson committed:
5.0.0-alpha.7

dougwilson commented:
Express.js 5.0.0-alpha.7 has been published which includes the initial support for Promises in the router. Middleware and handlers can now return promises and if the promise is rejected, `next(err)` will be called with `err` being the value of the rejection. The implementation is seeking feedback from real usage, and please open any feedback as a new issue, either in this issue tracker or in the `router` issue tracker.

I am currently working on writing up Express.js-specific documentation on this feature, but in the meantime, the documentation can be found in the `router` repository:

https://github.com/pillarjs/router/tree/v2.0.0-alpha.1#middleware

> The function can optionally return a Promise object. If a Promise object is returned from the function, the router will attach an onRejected callback using .then. If the promise is rejected, next will be called with the rejected value, or an error if the value is falsy.

crandmck commented:
Great to see this progress!  Since it's still alpha, I don't think we need to create a bunch of official docs, but I would think an alert and link on the home page would be appropriate.  LMK if I can help.

o-t-w commented:
Is there any vague guestimate of a release date for the stable version?

dougwilson commented:
Your comment is irrelevant, I'm sorry. If you want a stable Express.js just use the current version. If you want to use preview features, just use the alpha. The only difference between them is API stability between versions, not how usable it is for production. The implementation is seeking feedback from real usage, and please open any feedback as a new issue, either in this issue tracker or in the router issue tracker.

72lions commented:
Great progress and I'm looking forward to start using Express v5. Is there any plan to have TypeScript types part of the package?

I apologise if this is not the right place to have this discussion.

wesleytodd commented:
@72lions You can see our conversation on this topic here: https://github.com/expressjs/express/issues/2818

dougwilson committed:
Merge tag '4.17.1'

StarpTech commented:
Hi, many months have passed since the last commit or comment. Is it still worth to expect a new express version or should we put the effort into other frameworks? I saw your comment https://github.com/expressjs/express/pull/2237#issuecomment-435078982 and I fully agree with you but it's also really hard for me to be on track with the status quo. How can we move forward? Thanks!

dougwilson commented:
> Is it still worth to expect a new express version

Yes

> or should we put the effort into other frameworks?

That is something you would have to decide; I don't know enough about your work/projects to make such a decision for you.

> I saw your comment #2237 (comment) and I fully agree with you but it's also really hard for me to be on track with the status quo. How can we move forward?

Please open a new issue with your feedback as requested in that comment. The PR has gotten very long for conversations.

wesleytodd commented:
It was last January when we said we would shoot for a release early summer for this.  Now that it is January again, I would like to take the lead in pushing this release out the door by the end of the month.  I know that sounds like "well why could we do it now when we couldn't all last year".  And the answer really is that no one is paid to work on this project and we all are very busy.  So call it a new years resolution or what, but I want us to make this happen so we can unblock so much development on Express 6.0.  So what will it take to make this happen:

1. Agree on the Jan 31, 2020 goal
2. Create an updated list of todo items
3. Knock those out
4. Prepare a blog post announcing it and describing the update process
5. Publish `5.0`!
6. Start work on `6.0`

I really think that a few solid days work is all we are away from having this out the door, and I will be taking time this month to make it happen.  So, for step 2, here is what I think the real todo list is:

1. #3621 Just a simple PR to open then merge
2. #2432 I started this in https://github.com/expressjs/discussions/issues/82, and it is pretty simple to finish
3. Status `undefined`: #2795 #2797 #3111 #3137 #3143
4. https://github.com/pillarjs/router/pull/42 Merge this with its current state. The `path-to-regexp@6` can go with `express@6`
5. Remove the `alpha` tag on the 5.0 docs

Am I missing anything?  If that is it then I will start on these and the list at the top can just be updated to be more concise.

As for the announcement, we can work with the OpenJS foundation blog I think, they might even be able to help us write it.  If not I am happy to also knock that out once the rest is done.

dougwilson committed:
5.0.0-alpha.8

belovsandr committed:
docs: fix typo in history

closes #4286
closes #4327

dougwilson committed:
deps: router@2.0.0-beta.1

closes #4321

dougwilson committed:
Merge tag '4.17.2'

dougwilson committed:
deps: body-parser@2.0.0-beta.1

dougwilson committed:
deps: array-flatten@3.0.0

dougwilson committed:
Use mime-types for file to content type mapping

dougwilson committed:
Drop support for Node.js below 4

dougwilson committed:
deps: serve-static@2.0.0-beta.1

dougwilson committed:
deps: send@1.0.0-beta.1

dougwilson committed:
docs: fix entry in history

dougwilson committed:
5.0.0-beta.1

dougwilson committed:
Merge tag '4.17.3'

EvanHahn committed:
Use object with null prototype for settings closes #4835

EvanHahn committed:
Use object with null prototype for various app properties

`app.cache`, `app.engines`, and `app.settings` are now created with
`Object.create(null)` instead of `{}`.

This also updates a test to ensure that `app.locals` is created the same
way.

wesleytodd committed:
Merge tag '4.19.0' into 5.x

4.19.0

wesleytodd committed:
Merge tag '4.19.1' into 5.x

4.19.1

wesleytodd committed:
build: disable test for 4.x which is failing in v5

wesleytodd committed:
5.0.0-beta.2

wesleytodd committed:
Improved fix for open redirect allow list bypass

Co-authored-by: Jon Church <me@jonchurch.com>
Co-authored-by: Blake Embrey <hello@blakeembrey.com>

wesleytodd committed:
4.19.2

wesleytodd committed:
Merge tag '4.19.2' into 5.0

4.19.2

wesleytodd committed:
Merge branch '5.0' into 5.x

wesleytodd committed:
v5.0.0-beta.3

UlisesGascon committed:
ci: adopt Node@18 as the minimum supported version

UlisesGascon committed:
feat: adopt Node@18 as the minimum supported version

ctcpip committed:
💚 add legacy CI, clean up

ctcpip committed:
Merge branch 'master' into 5.0

jonchurch committed:
Throw on invalid status codes (#4212)

* check status code is integer, or string integer, in range

* fix tests, update jsdoc comment for res.status

* throw if number is string

* narrow valid range to between 1xx and 5xx

* disambiguate the error message

* update skipped tests, remove invalid string test

* remove invalid float test

* fixup! remove invalid float test

* fix invalid range tests error assertions

* remove unused deprecate function

* add test to assert on 200.00 coming through as 200

this is the behavior of node's underlying HTTP module

* revert back to throwing only on > 999 and < 100

* update implementation for > 999

* add test for 700 status code

* update history with change

* update jsdoc

* clarify jsdoc comment

* one more round of jsdoc

* update 501 test

* add invalid status code test for res.sendStatus

* add test describe block for valid range

* fixup! add test describe block for valid range

* reduce the describe nesting

* switch to testing status 100, to avoid 100-continue behavior

* fix 900 test

* stringify code in thrown RangeError message

* remove accidentally duplicated res.status method

* fix error range message

Co-authored-by: Chris de Almeida <ctcpip@users.noreply.github.com>

* update sendStatus invalid code test to use sendStatus

---------

Co-authored-by: Chris de Almeida <ctcpip@users.noreply.github.com>

almic committed:
Use Array.flat instead of array-flatten (#5677)

UlisesGascon committed:
feat: adopt Node@18 as the minimum supported version (#5803)

- PR-URL: https://github.com/expressjs/express/pull/5803
- This is a BREAKING CHANGE

ctcpip committed:
💚 remove node <11, all failing permanently now

jonchurch committed:
Ignore `expires` and `maxAge` in `res.clearCookie()` (#5792)

* add test for removing user provided expires

* rework impl and tests to ignore maxAge, do not set it

this is to take into account the built-in relative expires when passing
a maxAge to res.cookie

I realized that using maxAge to invalidate cookies inherrently hit this
relativee expires behavior, and the goal of this PR is not to rework
that relative expires behavior w/ maxAge, but to prevent users from
overwriting these values by accident when clearing cookies

* update history.md

* explicitly delete maxAge instead of setting as undefined

* drop the spread, use object.assign

* wording, review comment on history.md

Co-authored-by: Chris de Almeida <ctcpip@users.noreply.github.com>

* ♻️ use spread, update supported ecmascript version

---------

Co-authored-by: Chris de Almeida <ctcpip@users.noreply.github.com>

ctcpip committed:
Merge branch '5.0' into 5-merge

ctcpip committed:
Merge branch 'master' into 5-merge

ctcpip committed:
Merge branch 'ci/v5-node-lts' into 5-merge

wesleytodd committed:
fix(deps)!: send@^1.0.0 (#5786)

carpasse committed:
chore: upgrade `debug` dep from 3.10 to 4.3.6 (#5829)

* chore: upgrade debug dep from 3.10 to 4.3.6

* docs: add debug dep upgrade to History.md

carpasse committed:
refactor: replace 'path-is-absolute' dep with node:path isAbsolute method (#5830)

* refactor: replace 'path-is-absolute' dep with node:path isAbsolute method

* docs: add path-is-absolute dep removal to History.md

RobinTail committed:
5.x: Upgrading `merge-descriptors` with allowing minors (#5782)

* Upgrading `merge-descriptors` with allowing minors in v5

* Using ^2.0.0 as per request

* Reflecting in History.md.

* Update History.md

wesleytodd committed:
fix(deps) serve-staic@2.0.0 (#5790)

carpasse committed:
chore: qs@6.13.0 (#5847)

Co-authored-by: Wes Todd <wtodd@netflix.com>

IamLizu committed:
cookie-signature@^1.2.1 (#5833)

* upgraded `cookie-signature` to 1.2.1

* declared cookie-signature deps in history

* add caret in version

wesleytodd committed:
fix(deps): accepts@^2.0.0 (#5881)

wesleytodd committed:
fix(deps)!: mime-types@^3.0.0 (#5882)

wesleytodd committed:
fix(deps): type-is@^2.0.0 (#5883)

wesleytodd committed:
fix(deps)!: content-disposition@^1.0.0 (#5884)

wesleytodd committed:
fix(deps): finalhandler@^2.0.0 (#5899)

jonchurch committed:
update to fresh@2.0.0 (#5916)

fixes handling of If-Modified-Since in combination with If-None-Match

wesleytodd committed:
feat(deps)!: router@^2.0.0 (#5885)

wesleytodd committed:
Merge tag '4.20.0' into 5.0

wesleytodd committed:
Merge branch '5.0' into 5-merge

ctcpip committed:
🔧 update CI, remove unsupported versions, clean up

blakeembrey committed:
Delete `back` as a magic string (#5933)

wesleytodd committed:
fix(deps): body-parser@^2.0.1

wesleytodd committed:
fix(deps): serve-static@^2.1.0

wesleytodd committed:
fix(deps): send@^1.1.0

wesleytodd committed:
5.0.0

closes #2237
closes https://github.com/expressjs/discussions/issues/233

UlisesGascon reviewed (approved):
LGTM! Time to make it real!

karlhorky commented:
Thanks to all of the contributors for getting `express@5.0.0` published to the `next` tag on [npm](https://www.npmjs.com/package/express?activeTab=versions) 🚀 

![Screenshot 2024-09-10 at 10 10 28](https://github.com/user-attachments/assets/ef7d62e2-fb64-4148-933a-a7f3569ad8e7)

**Edit:** If you're interested in my points below, check out this issue that I opened:

- https://github.com/expressjs/express/issues/5944

Maybe I missed a v5 planning document somewhere - I'm looking for a few details to help determine how and when to migrate our projects:

1. When is it planned that `express@5.0.0` be published to the npm `latest` tag?
2. Is there a plan to publish an Express v4 -> v5 Migration Guide? (or a visible list of breaking changes with short suggestions about what to do about them) The closest thing that I can see is [this small `breaking` note in `History.md`](https://github.com/expressjs/express/blob/5.x/History.md#500--2024-09-10:~:text=path.isAbsolute%20instead-,breaking%3A,-res.status())
3. What's the current plan for the TypeScript types in DefinitelyTyped [`@types/express`](https://www.npmjs.com/package/@types/express) and [`@types/express-serve-static-core`](https://www.npmjs.com/package/@types/express-serve-static-core)? Does anything need to be changed here? Specifically, things like https://github.com/DefinitelyTyped/DefinitelyTyped/pull/69660#issuecomment-2179358937 cc DT maintainers @borisyankov @samijaber @NatoBoram @bombillazo @AndrewLeedham @gabritto @sandersn

DominusKelvin commented:
Great job @wesleytodd and everyone who worked on this issue.

wesleytodd commented:
Hey @karlhorky! Thanks for the thanks and the questions. Would you mind moving them into a new issue and we can edit this to link back there? I just dont want to make it even harder to find things burried at the end of a 10 year old issue. 🤣

EDIT: and I should add, this stuff is the main reason we are on `next` and not `latest`. There is a ton to figure out before folks get the new version by default or before we even ask folks to migrate. We want to ensure stability, and we would love folks to help with migration tooling and docs updates first.

lirantal commented:
Just dropped by to say this is very cool and thank you all for the hard work ❤️

karlhorky commented:
> Hey @karlhorky! Thanks for the thanks and the questions. Would you mind moving them into a new issue and we can edit this to link back there? I just dont want to make it even harder to find things burried at the end of a 10 year old issue. 🤣
> 
> EDIT: and I should add, this stuff is the main reason we are on `next` and not `latest`. There is a ton to figure out before folks get the new version by default or before we even ask folks to migrate. We want to ensure stability, and we would love folks to help with migration tooling and docs updates first.

Sounds good, I created a first issue over here:

- https://github.com/expressjs/express/issues/5944

Zertz commented:
I'd like to congratulate everyone involved in this release for their absolutely incredible perseverance in getting Express 5.0 to the finish line! Huge thanks for all your hard work, I'm really looking forward to upgrading! 🙏

bertolo1988 commented:
Congratulations for the hardwork 🎉
I followed this thread for years 😅

spperforce commented:
Thank you so much for releasing all these amazing stuff @wesleytodd . Do we have any plans to backport recent security patches (path-go-regexp, send etc.) in v4?

blakeembrey commented:
@spperforce As 4.x is still maintained, all security work has been done for that release. Is there something missing?

wobsoriano commented:
This was so cool to see. Congrats to everyone involved!

sandersn commented:
Here are the maintainers listed for the Definitely Typed packages:

### @types/express
- @borisyankov
- @CMUH
- @puneetar
- @dfrankland

### @types/express-serve-static-core
- @borisyankov
- @micksatana
- @JoseLion
- @dwrss
- @andoshin11

But the express types have only had a few changes since 2021, so I don't know how active its maintainers are.

wesleytodd commented:
If any of the folks in this list are interested in helping us refine our strategy for supporting typescript, please take a look at the new issue https://github.com/expressjs/express/issues/5944 to continue the discussion.

mcnaveen reviewed (commented):
LGTM 🫡
